# Lambda Functions

This type of functions allows you to write functions in a quick and potentially dirty way so I wouldn't advise you use them all the time but sometimes there are situations when they can come in a very handy . For example check the next example.

To declare a lambda function you have to specify the arguments of the function after of keyword **lambda**.

Compare lambda functions with convetional functions

In [14]:
# lambda way
sum1 = lambda x, y : x + y
sum1(3,4)


7

In [16]:
# conventional way
def sum2(x,y):
    return x+y
sum2(3,4)

7

In [20]:
%timeit sum1(100,50)

282 ns ± 88.5 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [21]:
%timeit sum2(100,50)

337 ns ± 109 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


**Exercise** : make the next conventional function one lambda function

In [23]:
def add_bangs1(word1):
    return word1 + "!!!"

add_bangs1('hello')

'hello!!!'

**Solution** :

In [12]:
add_bangs = (lambda a: a + '!!!')
add_bangs('hello')

'hello!!!'

### Anonymous functions:

## Map()

Function ´map()´ takes two arguments ´map( funct, seq). map() applies the function to All elements in the sequence. How the function, which will be applied to every value in the list (seq), doesn't has predefined name it will be one **anonymous function**.

The lambda function definition is: `add_bangs = (lambda a: a + '!!!')`, and the function call is: `add_bangs('hello')`.

In [11]:
x = [1,2,3,4,5,6]
squaredElements = map(lambda x : x**2, x)
print(list(squaredElements))

[1, 4, 9, 16, 25, 36]


In [25]:
def fahrenheit1(T):
    return ((float(9)/5)*T + 32)

def celsius1(T):
     return (float(5)/9)*(T-32)
 
temperatures = (36.5, 37, 37.5, 38, 39)
F = map(fahrenheit1, temperatures)
C = map(celsius, F)

temperatures_in_Fahrenheit = list(map(fahrenheit1, temperatures))
temperatures_in_Celsius = list(map(celsius1, temperatures_in_Fahrenheit))
print(temperatures_in_Fahrenheit)

print(temperatures_in_Celsius)


[97.7, 98.60000000000001, 99.5, 100.4, 102.2]
[36.5, 37.00000000000001, 37.5, 38.00000000000001, 39.0]


In [36]:
F1 = list(map(lambda T: (float(9)/5)*T + 32, temperatures))
C2 = list(map(lambda T: (float(5)/9)*(T-32), F1))
print(F1)
print(C2)

[97.7, 98.60000000000001, 99.5, 100.4, 102.2]
[36.5, 37.00000000000001, 37.5, 38.00000000000001, 39.0]


map() can be applied to more than one list. The lists don't have to have the same length. map() will apply its lambda function to the elements of the argument lists, i.e. it first applies to the elements with the 0th index, then to the elements with the 1st index until the n-th index is reached: 

In [82]:
a = [1, 2, 3, 4]
b = [17, 12, 11, 10]
c = [-1, -4, 5]
list(map(lambda x, y, z : 2.5*x + 2*y - z, a, b, c))

[37.5, 33.0, 24.5]

Note: If one list has fewer elements than the others, map will stop when the shortest list has been consumed

The map function of the previous chapter was used to apply one function to one or more iterables. We will now write a function which applies a bunch of functions, which may be an iterable such as a list or a tuple, for example, to one Python object.

In [67]:
from math import sin, cos, tan, pi
#import math also works
import numpy as np
def map_functions(x, functions):
    """ map an iterable of functions on the the object x """
    res=[]
    for i in x:
        for func in functions:
            res.append(func(i))
    return res

In [74]:
values = np.linspace(0,pi,100)
family_of_functions = (sin, cos, tan)
print(map_functions(values, family_of_functions))

[0.0, 1.0, 0.0, 0.03172793349806765, 0.9994965423831851, 0.03174391521396965, 0.0634239196565645, 0.9979866764718844, 0.06355187013195693, 0.09505604330418267, 0.9954719225730846, 0.09548842227361158, 0.12659245357374926, 0.9919548128307953, 0.12761917371264675, 0.1580013959733499, 0.9874388886763943, 0.16001131592572962, 0.18925124436041021, 0.9819286972627067, 0.19273420248128023, 0.22031053278654064, 0.975429786885407, 0.225859960141265, 0.2511479871810792, 0.9679487013963562, 0.2594641501339635, 0.28173255684142967, 0.9594929736144974, 0.29362649293836673, 0.3120334456984871, 0.9500711177409454, 0.32843167197886425, 0.3420201433256687, 0.9396926207859084, 0.36397023426620234, 0.3716624556603276, 0.9283679330160726, 0.4003396093754275, 0.4009305354066137, 0.9161084574320696, 0.437645272406344, 0.42979491208917164, 0.9029265382866212, 0.4760020819686433, 0.4582265217274104, 0.8888354486549235, 0.5155358310931968, 0.48619673610046865, 0.8738493770697849, 0.5563850577210417, 0.51367739

## Filtering
### Filter()

`filter(function, sequence)`

offers an elegant way to filter out all the elements of a sequence "sequence", for which the function function returns True. i.e. an item will be produced by the iterator result of filter(function, sequence) if item is included in the sequence "sequence" and if function(item) returns True. 

In other words: The function filter(f,l) needs a function f as its first argument. f has to return a Boolean value, i.e. either True or False. This function will be applied to every element of the list l. Only if f returns True will the element be produced by the iterator, which is the return value of filter(function, sequence). 


In [75]:
# Create a list of strings: fellowship
fellowship = ['frodo', 'samwise', 'merry', 'pippin', 'aragorn', 'boromir', 'legolas', 'gimli', 'gandalf']

# Use filter() to apply a lambda function over fellowship: result
result = filter(lambda a : len(a)>6, fellowship)

# Convert result to a list: result_list
result_list = list(result)


In the following example, we filter out first the odd and then the even elements of the sequence of the first 11 Fibonacci numbers: 

In [76]:
fibonacci = [0,1,1,2,3,5,8,13,21,34,55]
odd_numbers = list(filter(lambda x: x % 2, fibonacci))
print(odd_numbers)

even_numbers = list(filter(lambda x: x % 2 == 0, fibonacci))
print(even_numbers)

[1, 1, 3, 5, 13, 21, 55]
[0, 2, 8, 34]


## Reducing
### Reduce()

`reduce(func, seq)` 

continually applies the function func() to the sequence seq. It returns a single value. 

If seq = [ s1, s2, s3, ... , sn ], calling reduce(func, seq) works like this:

At first the first two elements of seq will be applied to func, i.e. func(s1,s2) The list on which reduce() works looks now like this: [ func(s1, s2), s3, ... , sn ]

In the next step func will be applied on the previous result and the third element of the list, i.e. func(func(s1, s2),s3)
The list looks like this now: [ func(func(s1, s2),s3), ... , sn ]

Continue like this until just one element is left and return this element as the result of reduce()


In [77]:
# Import reduce from functools
from functools import reduce

# Create a list of strings: stark
stark = ['robb', 'sansa', 'arya', 'brandon', 'rickon']

# Use reduce() to apply a lambda function over stark: result
result = reduce(lambda item1, item2 : item1 + item2, stark)

# Print the result

In [78]:
import functools
functools.reduce(lambda x,y: x+y, [47,11,42,13])

113

In [ ]:
from functools import reduce
f = lambda a,b: a if (a > b) else b
reduce(f, [47,11,42,102,13])
102

Imagine an accounting routine used in a book shop. It works on a list with sublists, which look like this: 

| Order Number | Book Title and Author | Quantity | Price per Item 
| ---------- | ---------- | ---------- | ---------- 
| 34587 |	Learning Python, Mark Lutz |	4 |	40.95
| 98762 | Programming Python, Mark Lutz |	5 |	56.80 
| 77226 |	Head First Python, Paul Barry |	3 |	32.95 
| 88112 |	Einführung in Python3, Bernd Klein |	3 |	24.99 

Write a Python program, which returns a list with 2-tuples. Each tuple consists of a the order number and the product of the price per items and the quantity. The product should be increased by 10,- € if the value of the order is smaller than 100,00 €. 
Write a Python program using lambda and map.


The same bookshop, but this time we work on a different list. The sublists of our lists look like this: 
[ordernumber, (article number, quantity, price per unit), ... (article number, quantity, price per unit) ] 
Write a program which returns a list of two tuples with (order number, total amount of order).

In [80]:
orders = [ ["34587", "Learning Python, Mark Lutz", 4, 40.95], 
	       ["98762", "Programming Python, Mark Lutz", 5, 56.80], 
           ["77226", "Head First Python, Paul Barry", 3,32.95],
           ["88112", "Einführung in Python3, Bernd Klein", 	3, 24.99]]

min_order = 100
invoice_totals = list(map(lambda x: x if x[1] >= min_order else (x[0], x[1] + 10), 
			              map(lambda x: (x[0],x[2] * x[3]), orders)))

print(invoice_totals)

[('34587', 163.8), ('98762', 284.0), ('77226', 108.85000000000001), ('88112', 84.97)]


The output of the previous program looks like this:
[('34587', 163.8), ('98762', 284.0), ('77226', 108.85000000000001), ('88112', 84.97)]

In [81]:
orders = [ [1, ("5464", 4, 9.99), ("8274",18,12.99), ("9744", 9, 44.95)], 
	       [2, ("5464", 9, 9.99), ("9744", 9, 44.95)],
	       [3, ("5464", 9, 9.99), ("88112", 11, 24.99)],
           [4, ("8732", 7, 11.99), ("7733",11,18.99), ("88112", 5, 39.95)] ]

min_order = 100
invoice_totals = list(map(lambda x: [x[0]] + list(map(lambda y: y[1]*y[2], x[1:])), orders))
invoice_totals = list(map(lambda x: [x[0]] + [reduce(lambda a,b: a + b, x[1:])], invoice_totals))
invoice_totals = list(map(lambda x: x if x[1] >= min_order else (x[0], x[1] + 10), invoice_totals))

print (invoice_totals)

[[1, 678.3299999999999], [2, 494.46000000000004], [3, 364.79999999999995], [4, 492.57]]


We will get the following result:
[[1, 678.3299999999999], [2, 494.46000000000004], [3, 364.79999999999995], [4, 492.57]]